cpp/src/arrow/util/decimal-benchmark.cc

```diff
@@ -45,53 +45,114 @@ static void FromString(benchmark::State& state) {  // NOLINT non-const reference
+      benchmark::DoNotOptimize(v1[x] == v2[x]);
+      benchmark::DoNotOptimize(v1[x + 1] <= v2[x + 1]);
+      benchmark::DoNotOptimize(v1[x + 2] >= v2[x + 2]);
+      benchmark::DoNotOptimize(v1[x + 3] >= v1[x + 3]);
+    }
+  }
+}
+
+static void BinaryCompareOpConstant(
+    benchmark::State& state) {  // NOLINT non-const reference
+  std::vector<BasicDecimal128> v1, v2;
+  for (int x = 0; x < 100; x++) {
+    v1.emplace_back(100 + x, 100 + x);
+    v2.emplace_back(200 + x, 200 + x);
+  }
+  for (auto _ : state) {
+    for (int x = 0; x < 25; x += 4) {
+      v2[x] = v1[x] == v1[0];
+      v2[x + 1] = v1[x + 1] <= v1[0];
+      v2[x + 2] = v1[x + 2] >= v1[0];
+      v2[x + 3] = v1[x + 3] != v1[0];
+    }
+  }
+}
+
+static void BinaryMathOpAggregate(
+    benchmark::State& state) {  // NOLINT non-const reference
+  std::vector<BasicDecimal128> v;
+  for (int x = 0; x < 100; x++) {
+    v.emplace_back(100 + x, 100 + x);
+  }
+
+  for (auto _ : state) {
+    BasicDecimal128 result;
+    for (int x = 0; x < 100; x += 5) {
+      result += v[x];
+    }
+    benchmark::DoNotOptimize(result);
   }
 }
 
 static void BinaryMathOp(benchmark::State& state) {  // NOLINT non-const reference
-  BasicDecimal128 d1(546, 123), d2(123, 456), d3(0, 10);
-  while (state.KeepRunning()) {
-    benchmark::DoNotOptimize(d1 - d2);
-    benchmark::DoNotOptimize(d1 + d2);
-    benchmark::DoNotOptimize(d1 * d2);
-    benchmark::DoNotOptimize(d1 / d2);
-    benchmark::DoNotOptimize(d1 % d3);
+  std::vector<BasicDecimal128> v1, v2;
+  for (int x = 0; x < 100; x++) {
+    v1.emplace_back(100 + x, 100 + x);
+    v2.emplace_back(200 + x, 200 + x);
+  }
+
+  for (auto _ : state) {
+    for (int x = 0; x < 100; x += 5) {
+      benchmark::DoNotOptimize(v1[x] - v2[x]);
+      benchmark::DoNotOptimize(v1[x + 1] + v2[x + 1]);
+      benchmark::DoNotOptimize(v1[x + 2] * v2[x + 2]);
+      benchmark::DoNotOptimize(v1[x + 3] / v2[x + 3]);
+      benchmark::DoNotOptimize(v1[x + 4] % v2[x + 4]);
+    }
   }
 }
 
 static void UnaryOp(benchmark::State& state) {  // NOLINT non-const reference
-  BasicDecimal128 d1(-546, 123), d2(-123, 456);
-  while (state.KeepRunning()) {
-    benchmark::DoNotOptimize(d1.Abs());
-    benchmark::DoNotOptimize(d2.Negate());
+  std::vector<BasicDecimal128> v;
+  for (int x = 0; x < 100; x++) {
+    v.emplace_back(100 + x, 100 + x);
+  }
+
+  for (auto _ : state) {
+    for (int x = 0; x < 100; x += 2) {
+      benchmark::DoNotOptimize(v[x].Abs());
+      benchmark::DoNotOptimize(v[x + 1].Negate());
+    }
   }
 }
 
 static void Constants(benchmark::State& state) {  // NOLINT non-const reference
   BasicDecimal128 d1(-546, 123), d2(-123, 456);
-  while (state.KeepRunning()) {
+  for (auto _ : state) {
     benchmark::DoNotOptimize(BasicDecimal128::GetMaxValue() - d1);
     benchmark::DoNotOptimize(BasicDecimal128::GetScaleMultiplier(3) + d2);
   }
 }
 
 static void BinaryBitOp(benchmark::State& state) {  // NOLINT non-const reference
```

Comment:
Wow, I didn't know we even had decimal bit ops. What's the point? @pravindra do you know?

Reply:
The &= and |= have always been there.